CRM/Core/Permission/Joomla.php

```diff
@@ -131,4 +135,74 @@ public function checkGroupRole($array) {
     return FALSE;
   }
 
+  /**
+   * @inheritDoc
+   */
+  public function upgradePermissions($permissions) {
+    $translatedPerms = array();
+    foreach (array_flip($permissions) as $perm) {
+      $translated = $this->translateJoomlaPermission($perm);
+      $translatedPerms[] = $translated[0];
+    }
+
+    $associations = $this->getUserGroupPermsAssociations();
+    $originalAssociations = clone $associations;
+    foreach (array_keys(get_object_vars($associations)) as $permName) {
+      if (!in_array($permName, $translatedPerms)) {
+        unset($associations->$permName);
+      }
+    }
+
+    if ($originalAssociations !== $associations) {
```

Comment:
By my reading of the linked documentation, '!=' (one less equals sign) would work here as I intended. Do you see a reason to use a flag instead?

For what it's worth, the consequences of getting this wrong are pretty low. I introduced the conditional as an optimization to avoid an unnecessary write to the database (i.e., don't write the permissions back if they haven't changed).

... which isn't to say we shouldn't get it right....